### PR TITLE
feat: add `TransitionDirection` and visibility method for `Transition`-class

### PR DESCRIPTION
**Pull request by simbig**
- [x] Added automated tests
- [x] Documented for all relevant versions
- [x] Updated the changelog

**Changes**

- add `TransitionDirection` and visibility method for `Transition`-class

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,13 @@ See [GitHub releases](https://github.com/mll-lab/laravel-utils/releases).
 
 ## Unreleased
 
+
```
Comment (spawnia) [inline CHANGELOG.md]:
```suggestion
```

File: composer.json [hunk 2]
```diff
@@ -20,6 +20,7 @@
   },
   "require": {
     "php": "^8.1",
+    "bensampo/laravel-enum": "^6.3",
```
Comment (spawnia) [inline composer.json]: Given this project requires PHP 8.1, we can use proper enums here.

File: src/ModelStates/IsStateManager.php [hunk 3]
```diff
@@ -30,4 +30,15 @@ public function getCanTransitionToAttribute(): SupportCollection
...
+
+    /**
+     * @return SupportCollection<array-key, Transition>
+     */
+    public function possibleTransitions(): SupportCollection
+    {
+        $stateable = $this->stateable;
+
+        return $stateable->stateClass()::config()
+            ->possibleNextTransitions($stateable);
```
Comment (spawnia) [inline src/ModelStates/IsStateManager.php:42]:
```suggestion
        return $stateable->stateClass()
            ::config()
            ->possibleNextTransitions($stateable);
```

File: src/ModelStates/StateConfig.php [hunk 4]
```diff
@@ -107,4 +108,21 @@ public function possibleNextStates(State $from): SupportCollection
         // @phpstan-ignore-next-line php-stan is not right here
         return new SupportCollection($next);
     }
+
+    /**
+     * Returns a possibly empty list of the next possible transitions for a given model with StateManager.
+     *
+     * @return SupportCollection<array-key, Transition>
```
Comment (spawnia) [inline src/ModelStates/StateConfig.php]:
```suggestion
     * @return SupportCollection<int, Transition>
```

File: src/ModelStates/StateConfig.php [hunk 5]
```diff
@@ -107,4 +108,21 @@ public function possibleNextStates(State $from): SupportCollection
...
+    public function possibleNextTransitions(Model&HasStateManagerInterface $stateable): SupportCollection
+    {
+        $stateable->refresh();
+        $from = $stateable->state;
+        $possibleNextStates = $this->possibleNextStates($from);
+
+        $stateMachine = $stateable->stateMachine();
+
+        return $possibleNextStates
+            ->map(fn (State $nextState): Transition => $stateMachine->instantiateTransitionClass($from::class, $nextState::class))->values();
```
Comment (spawnia) [inline src/ModelStates/StateConfig.php]:
```suggestion
            ->map(fn (State $nextState): Transition => $stateMachine->instantiateTransitionClass($from::class, $nextState::class))
            ->values();
```

File: src/ModelStates/StateMachine.php [hunk 6]
```diff
@@ -51,4 +46,14 @@ public function transitionTo(State|string $newState): HasStateManagerInterface
 
         return $this->model;
     }
+
+    public function instantiateTransitionClass(string $from, string $to): Transition
+    {
+        $transitionClass = $this->stateConfig->transition($from, $to);
+        if (null === $transitionClass) {
+            throw new TransitionNotFound($from, $to, $this->model);
+        }
```
Comment (spawnia) [inline src/ModelStates/StateMachine.php]:
```suggestion
        $transitionClass = $this->stateConfig->transition($from, $to)
            ?? throw new TransitionNotFound($from, $to, $this->model);
```

File: src/ModelStates/Transition.php [hunk 7]
```diff
@@ -21,6 +21,26 @@ public function canTransition(): bool
...
 
+    public function direction(): TransitionDirection
+    {
+        return TransitionDirection::FORWARD();
+    }
+
+    public function isVisibleFromUserFrontend(): bool
+    {
+        return true;
+    }
```
Comment (spawnia) [inline src/ModelStates/Transition.php]: UserFrontend as opposed to AdminFrontend? Can you explain in a comment?

File: src/ModelStates/Transition.php [hunk 8]
```diff
@@ -21,6 +21,26 @@ public function canTransition(): bool
...
+
+    public function to(): State
+    {
+        return new $this->to();
```
Comment (spawnia) [inline src/ModelStates/Transition.php:40]:
```suggestion
        return new $this->to;
```

or

```suggestion
        return (new $this->to)();
```

File: tests/ModelStates/StateTest.php [hunk 9]
```diff
@@ -113,6 +118,36 @@ public function testReturnsListOfPossibleNextStates(): void
...
 
+    public function testReturnsListOfPossibleNextTransitionsForCustomTransition(): void
+    {
+        $model = new TestModelWithCustomTransition();
+        $model->save();
+
+        $first = $model->stateManager->possibleTransitions()->first();
+        self::assertInstanceOf(CustomTransition::class, $first);
+        self::assertTrue($first->direction()->is(TransitionDirection::FORWARD));
+        self::assertSame(StateX::name(), $first->from()::name());
```
Comment (spawnia) [inline tests/ModelStates/StateTest.php]:
```suggestion
        self::assertInstanceOf(StateX::class, $first->from());
```

File: src/ModelStates/Transition.php (hunk 8 above)
Comment (simbig) [reply src/ModelStates/Transition.php:40]: `$this->to` is a `class-string<State> $to`. So to me, the instantiation looks fine... It also works fine as we can see in the test.

`self::assertSame(StateX::name(), $last->from()::name());`

Comment (spawnia) [reply src/ModelStates/Transition.php:40]: I just find the operator precedence really unexpected here, at first glance I though this is a recursive call.

File: src/ModelStates/Transition.php (hunk 7 above)
Comment (spawnia) [reply src/ModelStates/Transition.php]: Why not just `Frontend`? I was just confused by the term `UserFrontend`, as it implies there are multiple frontends.

**Review by spawnia — APPROVED** [review]

**Pull request merged**